java/arcs/android/storage/service/BindingContextStats.kt

```diff
@@ -69,4 +75,37 @@ class BindingContextStatsImpl : BindingContextStatistics {
             }
         }
     }
+
+    override fun traceTransaction(tag: String?, block: () -> Unit) {
+        // TODO(ianchang): Inject Android system traces with [tag]
+        ++_transactions
+        block()
+        --_transactions
+    }
+
+    companion object {
+        private var _transactions = Transactions(Mutex())
+    }
+}
+
+data class Transactions(val mutex: Mutex? = null, var current: Int = 0, var peak: Int = 0) {
+    operator fun inc(): Transactions {
+        return apply {
+            mutex?.let {
+                while (!mutex.tryLock()) {}
+                peak = maxOf(++current, peak)
+                mutex.unlock()
+            }
+        }
+    }
```

Comment:
If Transactions remains mutable, and these functions remain, this could be tidied up a bit:

```kotlin
    suspend operator fun inc() = apply {
        val incrementer = { peak = maxOf(current++, peak) }
        mutex?.withLock(incrementer) ?: incrementer()
    }
```
Also note: it's not generally a good practice to block on a mutex lock with a looping `tryLock`,  instead it's preferable to leverage `mutex.withLock`'s suspending capability.

(similarly for `dec`)

Reply:
I did this at the very first place but I changed it because I need the ++/-- also be synchronous for accurate tracing.
Which means I need them to be executed at its calling binder thread context rather than any coroutine context which potentially offsets the trace timestamp.